### PR TITLE
[REVIEW] read_csv support for user-specified decimal point and thousands separator 

### PR DESCRIPTION
**Pull request by vuule**
Implemented support for custom decimal point and thousands separator. 

Interface details:
* New parameters are optional, default decimal point is '.', default thousands separator is None.
* Decimal point cannot be the same as the delimiter.

Implementation details:
* Thousands separator is not applied to date/time parsing.
* Fixed the old behavior where comma was implicitly used as a thousands separator for both integers and floats. New default is consistent with pandas implementation.
* Added separate tests for the two options.
* Reduced the block size due to memory constraints - the added kernel parameters (decimal, thousands) caused failure with the original block size.

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

File: cpp/src/io/csv/csv_reader.cu [hunk 1]
```diff
@@ -255,6 +259,11 @@ gdf_error read_csv(csv_read_arg *args)
 	}
 
 	raw_csv->dayfirst = args->dayfirst;
+	raw_csv->decimal = args->decimal;
+	raw_csv->thousands = args->thousands == nullptr ? '\0' : *args->thousands;
+
+	if (raw_csv->decimal == raw_csv->delimiter){ checkError(GDF_INVALID_API_CALL, "Decimal point cannot be the same as the delimiter"); }
```
Comment (harrism) [inline cpp/src/io/csv/csv_reader.cu]: Please split long lines.

File: cpp/src/io/csv/csv_reader.cu [hunk 2]
```diff
@@ -87,6 +87,8 @@ typedef struct raw_csv_ {
     bool* 				d_parseCol;		// device : array of booleans stating if column should be parsed in reading process: parseCol[x]=false means that the column x needs to be filtered out.
     long 				header_row;		// Row id of the header
     bool				dayfirst;
+	char				decimal;
+	char				thousands;
 } raw_csv_t;
```
Comment (harrism) [inline cpp/src/io/csv/csv_reader.cu]: This file is obviously C++ (it uses std:: things), so it should define structs using C++ idiom: `struct Name { ... };`, NOT the C idiom `typedef struct Name_ { } Name;`

File: cpp/include/cudf/io_types.h [hunk 3]
```diff
@@ -75,9 +75,9 @@ typedef struct {
   bool			dayfirst;					// is the first value the day?  DD/MM  versus MM/DD
 
   char			*compression;				// specify the type of compression
-  char			*thousands;					// single character		a separate within numeric data  - if this matches the delimiter then system will return NULL
+  char			*thousands;					/**<  pointer to a single character that separates thousands in numeric data. If this matches the delimiter then system will return NULL */
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/io_types.h]: In fact, it looks like you indeed return `GDF_INVALID_API_CALL` in this case. Please update comment.

File: cpp/include/cudf/io_types.h [hunk 4]
```diff
@@ -75,9 +75,9 @@ typedef struct {
...
 
-  char			decimal;					// the decimal point character
+  char			decimal;			      /**< the decimal point character.If this matches the delimiter then system will return NULL */
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/io_types.h]: Please update comment to say it returns GDF_INVALID_API_CALL.

**Comment by harrism** [comment]
run tests

File: cpp/include/cudf/io_types.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (vuule) [reply cpp/include/cudf/io_types.h]: Are you sure about this change? /**< comment style is used for all other documented parameters

[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/io_types.h]: They didn't read our documentation style guide. :)

**Comment by harrism** [comment]
add to whitelist

**Pull request merged**